### PR TITLE
feat: add wsproxy implementation for key fetching

### PR DESCRIPTION
**Pull request by sreya**
This PR implements the `wsproxy` variant for a key cache. Along with `jwt` PR these are the final pieces before we can start wiring everything up.

### CODE REVIEW

File: enterprise/wsproxy/keycache.go [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+	if err != nil {
+		return codersdk.CryptoKey{}, xerrors.Errorf("fetch: %w", err)
+	}
+
+	if !k.latest.CanSign(now) {
+		return codersdk.CryptoKey{}, cryptokeys.ErrKeyNotFound
+	}
+
+	return k.latest, nil
+}
```
Comment (Emyrk) [inline enterprise/wsproxy/keycache.go:94]: This looks very similar to the `dbcache` [`dbkeycache.go`](https://github.com/coder/coder/blob/0aa84b18a10382f70bda7ee4b1c3f4b6bdcd22f9/coderd/cryptokeys/dbkeycache.go#L108-L142).

Can we DRY up these keycache implementations?

File: enterprise/wsproxy/keycache.go [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+	return k.closed.Load()
+}
+
+func (k *CryptoKeyCache) Close() {
+	if k.isClosed() {
+		return
+	}
+
+	k.refreshCancel()
+	k.closed.Store(true)
```
Comment (spikecurtis) [inline enterprise/wsproxy/keycache.go:241]: Multiple calls to `Close()` could race such that they both execute.  Right now `Close()` is idempotent, but I think it's better to just make `closed` a regular boolean protected by the mutex, so we don't have to worry about it getting executed more than once.

File: enterprise/wsproxy/keycache_test.go [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+		_, err = cache.Signing(ctx)
+		require.ErrorIs(t, err, cryptokeys.ErrClosed)
+
+		_, err = cache.Verifying(ctx, expected.Sequence)
+		require.ErrorIs(t, err, cryptokeys.ErrClosed)
+	})
+}
+
+type fakeCoderd struct {
+	server *httptest.Server
```
Comment (spikecurtis) [inline enterprise/wsproxy/keycache_test.go:467]: If you made `CryptoKeyCache.client` an interface, you wouldn't need to mess around with HTTP, URLs, or JSON.  You could just directly return the keys you want to test with.

File: enterprise/wsproxy/keycache.go [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+		return nil, codersdk.CryptoKey{}, xerrors.Errorf("fetch keys: %w", err)
+	}
+
+	if len(keys) == 0 {
+		return nil, codersdk.CryptoKey{}, cryptokeys.ErrKeyNotFound
+	}
+
+	now := k.Clock.Now()
+	if !latest.CanSign(now) {
+		return nil, codersdk.CryptoKey{}, cryptokeys.ErrKeyInvalid
```
Comment (spikecurtis) [inline enterprise/wsproxy/keycache.go:198]: Rare, but if you're fetching to get a verification key, it seems wrong to throw an error based on the signing key.

File: enterprise/wsproxy/keycache.go [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+	}
+
+	k.keysMu.RLock()
+	lastFetch := k.lastFetch
+	k.keysMu.RUnlock()
+
+	// There's a window we must account for where the timer fires while a fetch
+	// is ongoing but prior to the timer getting reset. In this case we want to
+	// avoid double fetching.
+	if now.Sub(lastFetch) < time.Minute*10 {
```
Comment (spikecurtis) [inline enterprise/wsproxy/keycache.go:161]: pull 10m into a constant

**Comment by Emyrk** [comment]
Spike's comments I think highlight the need to implement this once. With the changes https://github.com/coder/coder/pull/14928, can we reuse 1 implementation?

**Comment by sreya** [comment]
@Emyrk I'm game to refactor this to use one implementation but honestly I'd like to get this initiative over the finish line. I'd consider the refactor tech debt which I can tackle concurrently to ensuring there are no larger issues with the overall feature while we dogfood it in dev. You chill with that?

**Comment by Emyrk** [comment]
> @Emyrk I'm game to refactor this to use one implementation but honestly I'd like to get this initiative over the finish line. I'd consider the refactor tech debt which I can tackle concurrently to ensuring there are no larger issues with the overall feature while we dogfood it in dev. You chill with that?

I can live with it :+1:

File: enterprise/wsproxy/keycache.go [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+	refreshCtx    context.Context
+	refreshCancel context.CancelFunc
+	client        *wsproxysdk.Client
+	logger        slog.Logger
+	Clock         quartz.Clock
+
+	keysMu    sync.RWMutex
+	keys      map[int32]codersdk.CryptoKey
+	latest    codersdk.CryptoKey
+	fetchLock sync.RWMutex
```
[the remark this answers is not in the document]
Comment (sreya) [reply enterprise/wsproxy/keycache.go:32]: Honestly I haven't messed with sync.Cond that much this is super clean

File: enterprise/wsproxy/keycache.go [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+	key, ok = k.key(sequence)
+	if !ok {
+		return codersdk.CryptoKey{}, cryptokeys.ErrKeyNotFound
+	}
+
+	return checkKey(key, sequence, k.Clock.Now())
+}
+
+func (k *CryptoKeyCache) key(sequence int32) (codersdk.CryptoKey, bool) {
+	if sequence == latestSequence {
```
Comment (spikecurtis) [inline enterprise/wsproxy/keycache.go:124]: if you're going to use a special int32 to represent latest, you might as well go whole hog and just insert the crypto key into the map under that special value, and drop `k.latest`

**Review by spikecurtis — APPROVED** [review]

**Pull request merged**